sw/device/tests/sim_dv/i2c_device_tx_rx_test.c

```diff
@@ -0,0 +1,225 @@
+
+  // Temporary hack that connects i2c to a couple of open drain pins.
+  CHECK_DIF_OK(dif_pinmux_input_select(&pinmux,
+                                       kTopEarlgreyPinmuxPeripheralInI2c0Scl,
+                                       kTopEarlgreyPinmuxInselIob11));
+  CHECK_DIF_OK(dif_pinmux_input_select(&pinmux,
+                                       kTopEarlgreyPinmuxPeripheralInI2c0Sda,
+                                       kTopEarlgreyPinmuxInselIob12));
+  CHECK_DIF_OK(dif_pinmux_output_select(&pinmux, kTopEarlgreyPinmuxMioOutIob11,
+                                        kTopEarlgreyPinmuxOutselI2c0Scl));
+  CHECK_DIF_OK(dif_pinmux_output_select(&pinmux, kTopEarlgreyPinmuxMioOutIob12,
+                                        kTopEarlgreyPinmuxOutselI2c0Sda));
+
+  // I2C speed parameters.
+  dif_i2c_timing_config_t timing_config = {
+      .lowest_target_device_speed = kDifI2cSpeedFastPlus,
+      .clock_period_nanos = kClockPeriodNanos,
+      .sda_rise_nanos = kI2cRiseFallNanos,
+      .sda_fall_nanos = kI2cRiseFallNanos,
+      .scl_period_nanos = kI2cClockPeriodNanos};
+
+  dif_i2c_config_t config;
+  CHECK_DIF_OK(dif_i2c_compute_timing(timing_config, &config));
+  CHECK_DIF_OK(dif_i2c_configure(&i2c, config));
+  dif_i2c_id_t id0 = {.mask = kI2cDeviceMask0, .address = kI2cDeviceAddress0},
+               id1 = {.mask = kI2cDeviceMask1, .address = kI2cDeviceAddress1};
+  CHECK_DIF_OK(dif_i2c_set_device_id(&i2c, &id0, &id1));
+  CHECK_DIF_OK(dif_i2c_device_set_enabled(&i2c, kDifToggleEnabled));
+  CHECK_DIF_OK(
+      dif_i2c_irq_set_enabled(&i2c, kDifI2cIrqTxEmpty, kDifToggleEnabled));
+  CHECK_DIF_OK(dif_i2c_irq_set_enabled(&i2c, kDifI2cIrqTransComplete,
+                                       kDifToggleEnabled));
+
+  // Randomize variables.
+  uint8_t expected_data[kI2cByteCount];
+  LOG_INFO("Loopback %d bytes with addresses %0h, %0h", kI2cByteCount,
+           kI2cDeviceAddress0, kI2cDeviceAddress1);
+
+  // Controlling the randomization from C side is a bit slow, but might be
+  // easier for portability to a different setup later.
+  for (uint32_t i = 0; i < kI2cByteCount; ++i) {
+    expected_data[i] = rand_testutils_gen32_range(0, 0xff);
+  };
+
+  // Write expected data to transmit register.
+  CHECK(!tx_empty_irq_seen);
+  for (uint32_t i = 0; i < kI2cByteCount; ++i) {
+    CHECK_DIF_OK(dif_i2c_transmit_byte(&i2c, expected_data[i]));
+  }
+
+  uint8_t tx_fifo_lvl, acq_fifo_lvl;
+
+  // Make sure all fifo entries have been drained.
+  do {
+    CHECK_DIF_OK(
+        dif_i2c_get_fifo_levels(&i2c, NULL, NULL, &tx_fifo_lvl, &acq_fifo_lvl));
+  } while (tx_fifo_lvl > 0);
+  CHECK(tx_empty_irq_seen);
+
+  // Read data from i2c device.
+  CHECK(!trans_complete_irq_seen);
+
+  // Make sure all data has been read back.
+  do {
+    CHECK_DIF_OK(
+        dif_i2c_get_fifo_levels(&i2c, NULL, NULL, &tx_fifo_lvl, &acq_fifo_lvl));
+  } while (acq_fifo_lvl < kI2cByteCount + 2);  // acquired message, address and
```

Comment:
acq_fifo should also have a record of the initial read and stop

Reply:
Adding that should involve code reuse and probably some functions in i2c_testutils for the device responding to reads and writes.